qa/suites/orch/cephadm/workunits/task/test_rgw_multisite.yaml

```diff
@@ -16,6 +16,7 @@ tasks:
 - cephadm.shell:
     host.a:
       - ceph mgr module enable rgw
+      - for i in {1..5}; do TOKEN=$(./bin/ceph rgw realm tokens || true); if [ "$TOKEN" = "[]" ]; then break; fi; sleep 10; done
```

Comment:
- Why is `./bin/` prefix needed?  `ceph mgr module enable rgw` command doesn't have it.
- Is sleeping for 10 seconds justified?  Just checking that we wouldn't be unnecessarily delaying this test by 9 seconds on average.
- The `|| true` bit make me think that a simpler `while ! ceph rgw realm tokens; do sleep XYZ; done` loop could do.

Reply:
But honestly, I'd much rather see this addressed in `ceph mgr module enable` itself than play whack-a-mole with similar tests.  Is there really no condition we could block `ceph mgr module enable` on?

Reply:
> * Why is `./bin/` prefix needed?  `ceph mgr module enable rgw` command doesn't have it.

It's not. I will remove it.

> * Is sleeping for 10 seconds justified?  Just checking that we wouldn't be unnecessarily delaying this test by 9 seconds on average.

Test run times of the my 3 passing runs and the 2 passing runs you'd linked seem to be at least 15 minutes. Are you ok with the sleeping for 5 seconds, and retrying 24 times, i.e., 2 min?

> * The `|| true` bit make me think that a simpler `while ! ceph rgw realm tokens; do sleep XYZ; done` loop could do.

I wanted to introduce an explicit time limit for checking whether the module is loaded. I'm not sure what the default timeout set by teuthology is?

Reply:
> But honestly, I'd much rather see this addressed in `ceph mgr module enable` itself than play whack-a-mole with similar tests. Is there really no condition we could block `ceph mgr module enable` command on?

`ceph mgr module enable` adds a module to the enabled set of modules in the MgrMap.  When the mgr receives the updated MgrMap, respawns and loads the modules in the mgr's finisher queue. I guess when the set of pending module to be loaded is empty in Mgr (https://github.com/ceph/ceph/pull/35760/commits/2d9b3abd1fc50e5fcd9ce2c05e8fac41d389b052#diff-e70f87e77a44a4224ac1a612eae12ec0e5bad731e0f0c5b4eb4ce139c9858457R42), we can have the mgr indicate to the mon that all the modules have been loaded via the MgrMap? Then the monitor call blocks until it receives the mgrmap indicating that all the modules have been loaded? This doesn't seem trivial to do. Currently, the mgr sends messages to the monitor if a module has failed to load or has unmet dependencies.

IMO, the tests that use `mgr module enable` and immediately fire off mgr module commands should be fixed. They should check that the mgr is actually loaded (and not just enabled) before issuing the mgr command.

Reply:
> Test run times of the my 3 passing runs and the 2 passing runs you'd linked seem to be at least 15 minutes. Are you ok with the sleeping for 5 seconds, and retrying 24 times, i.e., 2 min?

I don't think we need a limit on the number retries: this is expected to succeed and we don't place timeouts on each individual `ceph` command, do we?  Broadly speaking, any command can hang and the overall teuthology job timeout is expected to take care of that.  But this is just a nit, feel free to ignore.

Reply:
> IMO, the tests that use `mgr module enable` and immediately fire off mgr module commands should be fixed. They should check that the mgr is actually loaded (and not just enabled) before issuing the mgr command.

I assume you meant "should check that the module is actually loaded (and not just enabled) before issuing the mgr command" command above.

We can certainly fix the tests -- I'm more concerned about the user experience.  If someone is automating something with a script, they would need to do roughly what you are doing here: identify some kind of info or status-reporting command provided by the module, try it and use the exit code to tell whether the module is loaded.  This is going to be different for every module and not intuitive at all...

Reply:
> > Test run times of the my 3 passing runs and the 2 passing runs you'd linked seem to be at least 15 minutes. Are you ok with the sleeping for 5 seconds, and retrying 24 times, i.e., 2 min?
> 
> I don't think we need a limit on the number retries: this is expected to succeed and we don't place timeouts on each individual `ceph` command, do we? Broadly speaking, any command can hang and the overall teuthology job timeout is expected to take care of that. But this is just a nit, feel free to ignore.

Done.

Reply:
> > IMO, the tests that use `mgr module enable` and immediately fire off mgr module commands should be fixed. They should check that the mgr is actually loaded (and not just enabled) before issuing the mgr command.
> 
> I assume you meant "should check that the _module_ is actually loaded (and not just enabled) before issuing the mgr command" command above.
> 
> We can certainly fix the tests -- I'm more concerned about the user experience. If someone is automating something with a script, they would need to do roughly what you are doing here: identify some kind of info or status-reporting command provided by the module, try it and use the exit code to tell whether the module is actually loaded. This is going to be different for every module and not intuitive at all...
> 
> But hopefully modules don't get enabled often, making this "fine" or at least acceptable?

For now, I added a note in doc/mgr about `mgr enable` not immediately loading a module. let me know if it's sufficient.

Reply:
@

> > > IMO, the tests that use `mgr module enable` and immediately fire off mgr module commands should be fixed. They should check that the mgr is actually loaded (and not just enabled) before issuing the mgr command.
> > 
> > 
> > I assume you meant "should check that the _module_ is actually loaded (and not just enabled) before issuing the mgr command" command above.
> > We can certainly fix the tests -- I'm more concerned about the user experience. If someone is automating something with a script, they would need to do roughly what you are doing here: identify some kind of info or status-reporting command provided by the module, try it and use the exit code to tell whether the module is actually loaded. This is going to be different for every module and not intuitive at all...
> > But hopefully modules don't get enabled often, making this "fine" or at least acceptable?
> 
> For now, I added a note in doc/mgr about `mgr enable` not immediately loading a module. let me know if it's sufficient.

@idryomov PTAL when you've time